### ISSUE TITLE
Support for ISO/IEC full subfigures (ISO DIR 2:2021 28.3.2) (was: Subfigures not working if Key is included)

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/metanorma-iso#754]
(UPDATED: 20220731)

## ISO/IEC DIR 2 supports Figure with subfigures

ISO/IEC DIR 2, 28.3.2 describes "subfigures"[here](https://www.iso.org/sites/directives/current/part2/index.xhtml). This feature has been in DIR 2 since at least 2011.

A figure that contains subfigures supports the following elements:
- a statement concerning units
- multiple subfigures
- a key for the subfigures, as a definition list
- additional paragraphs that describe the subfigures (normal clause text with NOTEs, but without tables, figures, examples)
- footnotes (which are referred to by content in the subfigure captions, paragraphs and NOTEs within the figure)

## Metanorma current support of subfigures

Currently in Metanorma [subfigures are supported in ISO](https://www.metanorma.org/author/topics/document-format/text/) but only the simplest form, of multiple images/figures.

<img width="791" alt="image" src="https://user-images.githubusercontent.com/11865/179803748-d8c23d0d-dbb1-46ca-8308-8eee0c3897c9.png">

Here, it describes that the "Key" content exists within the Figure.

Note that the subfigure example we have currently lacks a few things:
1. a "Statement concerning units" before the first subfigure.
2. a "Key" container that has paragraphs/lists after the last subfigure
3. Footnotes to the figure.

Metanorma also supports a figure with non-images:
<img width="774" alt="Screen Shot 2022-07-31 at 2 35 29 PM" src="https://user-images.githubusercontent.com/11865/182011823-7da33ad7-005a-49db-88ee-8647f5cb7e47.png">

This is the intended layout:

<img width="593" alt="Screen Shot 2022-07-31 at 3 11 26 PM" src="https://user-images.githubusercontent.com/11865/182012742-f4a97999-ca6f-4aae-a40b-6d0b656f68e2.png">

However, including the "Key" in the Figure doesn't work in Word, and PDF generation crashes.

```adoc
[[fig-C17-2]]
.Use of horizontal pattern and strata
====
.Grass with scattered patches of trees
image::img-C17-2a.png[]

.Grass with trees and grass without trees composed of 3-dimensional elementary features
image::img-C17-2b.png[]

.LCML horizontal pattern
image::img-C17-2c.png[]

*Key*:

This is an illustration of the approach of using horizontal pattern and
strata.

. The portion of the grass beneath the trees has different plant composition and
characteristics (density, height).

. The grass with trees and grass without trees, are treated as separate physical
features composed by 3-dimensional elementary features (tegons). This results in an
intrinsic (functional entity) mix between two types of land cover features having
different strata,

. It is described in LCML with the "horizontal pattern".
====
```

It becomes an EXAMPLE with 3 inner Figures, which is wrong:
<img width="502" alt="Screenshot 2022-07-20 at 1 40 36 AM" src="https://user-images.githubusercontent.com/11865/179804008-f2020ea4-252b-4f65-980e-5c6e5cfa6d50.png">

If I move the "Key" outside of the Figure, it works:
```adoc
[[fig-C17-2]]
.Use of horizontal pattern and strata
====
.Grass with scattered patches of trees
image::img-C17-2a.png[]

.Grass with trees and grass without trees composed of 3-dimensional elementary features
image::img-C17-2b.png[]

.LCML horizontal pattern
image::img-C17-2c.png[]

====

*Key*:

This is an illustration of the approach of using horizontal pattern and
strata.

. The portion of the grass beneath the trees has different plant composition and
characteristics (density, height).

. The grass with trees and grass without trees, are treated as separate physical
features composed by 3-dimensional elementary features (tegons). This results in an
intrinsic (functional entity) mix between two types of land cover features having
different strata,

. It is described in LCML with the "horizontal pattern".
```

<img width="486" alt="Screenshot 2022-07-20 at 1 44 12 AM" src="https://user-images.githubusercontent.com/11865/179804702-b3b8b25f-d326-42f1-bba6-cf4e3dfdca68.png">

## Metanorma with full subfigure support

We need to implement a complete `[figure]` block syntax to support figures with subfigures and their associated content.

Some functionality needed:
- Cross referencing from within the text in the figure to the subfigures
- Full `[figure]` syntax
- Recognition and rendering of `units` statement 
- Recognition and rendering of `key`
- Rendering of footnotes within the figure
- Explicit and implicit 

### Full figure syntax

```
[figure]
====
[%units]
A statement concerning units. footnote:[Footnote]

// explicit subfigure
[figure]
[[subfigure-a]]
.An explicit subfigure "a)"
======
image::rice_images/rice_image1.png[alt text]

Some content.
======

// implicit subfigure
.An implicit subfigure "b)"
image::rice_images/rice_image2.png[alt text]

[%key]
stem:[w]:: mass fraction of gelatinized kernels, expressed in per cent
stem:[t]:: cooking time, expressed in minutes
stem:[t_90]:: time required to gelatinize 90 % of the kernels footnote:[Another footnote]
P:: point of the curve corresponding to a cooking time of stem:[t_90]

Some paragraph text. <<subfigure-a>> describes...

NOTE: These results are based on a study carried out on three different types of kernel. footnote:[More footnotes]
====
```

While this is the full syntax for a figure/subfigures, there are "convenience syntaxes" for simpler figure inclusion.

### Figure with multiple images only => auto subfigures

We have this now:

```
[figure]
====
.An explicit subfigure "a)"
image::rice_images/rice_image1.png[alt text]

.An implicit subfigure "b)"
image::rice_images/rice_image2.png[alt text]
====
```

### Figure with image + text => one figure

We have this now:

```
[figure]
====
image::rice_images/rice_image1.png[alt text]

Some text.
====
```

### Figure with key after => one figure with key

We support this.

```
[figure]
.Some caption
====
image::rice_images/rice_image1.png[alt text]
====

[%key]
one:: one
two:: two
```

### Image with key after => one figure with key

We support this.

```
.Some caption
image::rice_images/rice_image1.png[alt text]

[%key]
one:: one
two:: two
```

### Figure with subfigures with key after => multiple figures with key

We don't support this.

```
[figure]
====
.An explicit subfigure "a)"
image::rice_images/rice_image1.png[alt text]

.An explicit subfigure "b)"
image::rice_images/rice_image2.png[alt text]
====

[%key]
one:: one
two:: two
```

Once this task is done we should also update the Metanorma site to provide a full example of subfigure with text/paragraphs.

NOTE: This is a blocker to https://github.com/metanorma/iso-19144-2.

**Comment by opoudjis** [comment]
Subfigures are recognised as examples containing only figures. 

Figure keys are supposed to be definition lists, and are recognised as such.

DISCURSIVE PARAGRAPHS ARE NOT DEFINITION LISTS. 

This is the first time I've seen any indication that anything other than definition lists is supposed to be a figure key. Every single other example in ISO/IEC DIR 2 of a table or figure key is a definition list, and the cited subfigure "key" is an umbrella description of notes and footnotes.

I take very strong exception to being told that I should have magically anticipated random markup that does not correspond to the common sense understanding of ISO/IEC DIR 2. This is not a key as far as Metanorma is concerned, and it is not encompassed by Metanorma syntax. I **REFUSE** to implement this without **EXPLICIT** feedback from ISO editorial that this is a legitimate instance of a key. As it is, if I were to implement this, it would NOT be with that existing markup: I would need to introduce a new open block syntax, and a much looser syntax.

I *VEHEMENTLY* recommend you do not disrupt Metanorma parsing and syntax over addle-headed frivolous use of ISO formatting. This is not a key. This is a note. And it should be treated as such.

**Comment by ronaldtse** [comment]
> Figure keys are supposed to be definition lists, and are recognised as such.

The image from ISO DIR 2 explicitly describes the "key" to contain "paragraphs".

> This is not a key. This is a note.

ISO DIR 2 says: 
> Separate keys, notes and footnotes for subfigures are not permitted

> I REFUSE to implement this without EXPLICIT feedback from ISO editorial that this is a legitimate instance of a key. 

Sorry, but ISO editorial are bound by ISO DIR 2. When ISO DIR 2 says it is the correct form, it is the correct form.

> if I were to implement this, it would NOT be with that existing markup: I would need to introduce a new open block syntax, and a much looser syntax.

I don't see why we cannot reuse the existing `====` block syntax with a role of "figure".

From the above reply I am not confident whether you have read the image:
![](https://user-images.githubusercontent.com/11865/179803911-869bd2c4-0a5e-40fd-b217-c1eaf5e14f52.png)

**Comment by opoudjis** [comment]
We are seeking confirmation on whether keys can be free text. The image is not a definition of keys, and it contradicts the definition of keys for tables. ISO DIR 2 is not explicit that the foregoing is a definition of keys; in fact, in its source diagram (in IEC, keys are mutually exclusive from requirement paragraphs and notes.

It does turn out that definition lists will incorrectly be embedded within subfigures (as nearest neighbours). That needs to be prevented, as it violates ISO/IEC DIR 2.

**Comment by ronaldtse** [comment]
I have updated the original ticket to update our current understanding. Will also update this once we receive feedback from ISO Editorial.

**Comment by ronaldtse** [comment]
I think we know what Keys are, they are the ones shown in the Rice document, which is not free text.

> It does turn out that definition lists will incorrectly be embedded within subfigures (as nearest neighbours). That needs to be prevented, as it violates ISO/IEC DIR 2.

I think we should stop merging the nearest definition list into the figure, it is very confusing for the author, especially when the definition list often is not part of the figure.

**Comment by opoudjis** [comment]
In fact we now also support this:

````
[figure]
====
.An explicit subfigure "a)"
image::rice_images/rice_image1.png[alt text]

.An explicit subfigure "b)"
image::rice_images/rice_image2.png[alt text]

[%key]
one:: one
two:: two

NOTE: Note
====
````

In fact, so will:

````
[figure]
====

.An explicit subfigure "a)"
=====
image::spec/assets/rice_image1.png[alt text]
=====

.An explicit subfigure "b)"
=====
image::spec/assets/rice_image1.png[alt text]
=====

[%key]
one:: one
two:: two

NOTE: Note
====
````

At issue is whether we allow `[%key]` to mark up open blocks of text as well.

> I think we should stop merging the nearest definition list into the figure, it is very confusing for the author, especially when the definition list often is not part of the figure.

We aren't: we are only merging definition lists marked with `[%key]` — signalling explicitly that they are associated with a figure or table. 

We also have the original syntax 

````

*Key*

a:: b
````

But I'm happy for that to be dropped from documentation, and kept only as legacy.

**Comment by ronaldtse** [comment]
> At issue is whether we allow [%key] to mark up open blocks of text as well.

I would say we should only allow text blocks in the figure if it was explicitly put in the figure block. 

I don’t think DIR 2 specified exactly what a “Key” is. 

> `*Key*` syntax  

Let’s make this legacy. And keep it as a note in documentation. 

BTW, does the Key syntax work the same way for Formulas?

**Comment by opoudjis** [comment]
> I would say we should only allow text blocks in the figure if it was explicitly put in the figure block.

That's not answering my question. Should text blocks be recognised *as keys*? That's your pending query to ISO.

Formulas work the same way: "where", or [%key], and again, what follows is expected to be a definition list.

**Comment by ronaldtse** [comment]
> Should text blocks be recognised as keys? That's your pending query to ISO.

<img width="323" alt="Screen Shot 2022-10-12 at 11 21 20 AM" src="https://user-images.githubusercontent.com/11865/195242254-34574622-ef11-4cda-9721-e883f570889e.png">

According to the figure, yes generic paragraphs (containing requirements) can be a key. This means that it would be better if the "key" is specified as an open block.

**Comment by opoudjis** [comment]
Units notes implemented: https://github.com/metanorma/metanorma-standoc/issues/688

**Comment by leo-barnes** [comment]
Are single-line subfigures like the one shown in the image [here](https://github.com/metanorma/metanorma-iso/issues/754#issuecomment-1200336176) supported?

In other words, can I somehow tell metanorma to _not_ place the images on separate lines?

**Comment by ronaldtse** [comment]
ISO/IEC DIR 2 has been updated with a Figure 5 that shows a Key with a paragraph.

<img width="938" height="631" alt="Image" src="https://github.com/user-attachments/assets/e78431d9-6db0-45eb-9bf3-d33c3cfce073" />

This tells me that Key needs to be in block syntax, it is not necessarily a definition list.

**Comment by opoudjis** [comment]
Well, that's dreadful. But of course, ISO don't author semantic models, they author Word documents.

Keys until now have been strictly definition lists, and manipulated as definition lists, with the definitions parsed as anchored to the figure or table, or whatever else.

Keys, because of the lax attitude of SDOs (who after all have only ever authored human-readable documents), can now indeed be discursive text. So instead of expecting to find `dl[@key = 'true']` right under the table or figure, we need to allow associated free text before and/or after, or instead of the definition list, and a container that attributes all that content as the key.

The least disruptive way of implementing this is to replace `dl` in the grammar with `key`, where `key = (p | dl)+`, and the `dl` has `@key = 'true'`, while `p` has `@type = 'key'`.

This is a grammar change, so I'm going to post it on metanorma-model-iso, and wait for a fortnight.
